src/LuaColor.cpp

```diff
@@ -0,0 +1,213 @@
+// Copyright © 2008-2019 Pioneer Developers. See AUTHORS.txt for details
+// Licensed under the terms of the GPL v3. See licenses/GPL-3.txt
+
+#include "LuaColor.h"
+#include "LuaUtils.h"
+#include "libs.h"
+
+inline void sanity_check(float r, float g, float b, float a)
+{
+	r = std::max(std::min(r, float(255.0)), float(0.0));
+	g = std::max(std::min(g, float(255.0)), float(0.0));
+	b = std::max(std::min(b, float(255.0)), float(0.0));
+	a = std::max(std::min(a, float(255.0)), float(0.0));
+}
+
+static int l_color_new(lua_State *L)
+{
+	LUA_DEBUG_START(L);
+	double r = luaL_checknumber(L, 1);
+	double g = luaL_checknumber(L, 2);
+	double b = luaL_checknumber(L, 3);
+	double a;
+	if (lua_isnumber(L, 4)) a = luaL_checknumber(L, 4);
+	else a = 255.0;
+	sanity_check(r, g, b, a);
+	LuaColor::PushToLua(L, Color(r, g, b, a));
+	LUA_DEBUG_END(L, 1);
+	return 1;
+}
+
+static int l_color_call(lua_State *L)
+{
+	LUA_DEBUG_START(L);
+	double r = luaL_checknumber(L, 2);
+	double g = luaL_checknumber(L, 3);
+	double b = luaL_checknumber(L, 4);
+	double a;
+	if (lua_isnumber(L, 5)) a = luaL_checknumber(L, 5);
+	else a = 255.0;
+	sanity_check(r, g, b, a);
+	Color c = Color(r, g, b, a);
+	LuaColor::PushToLua(L, Color(r, g, b, a));
+	LUA_DEBUG_END(L, 1);
+	return 1;
+}
+
+static int l_color_tostring(lua_State *L)
+{
+	const Color4ub *c = LuaColor::CheckFromLua(L, 1);
+	luaL_Buffer buf;
+	luaL_buffinit(L, &buf);
+	char *bufstr = luaL_prepbuffer(&buf);
+	int len = snprintf(bufstr, LUAL_BUFFERSIZE, "Color(%i, %i, %i, %i)", c->r, c->g, c->b, c->a);
+	assert(len < LUAL_BUFFERSIZE); // XXX should handle this condition more gracefully
+	luaL_addsize(&buf, len);
+	luaL_pushresult(&buf);
+	return 1;
+}
+
+static int l_color_add(lua_State *L)
+{
+	const Color4ub *a = LuaColor::CheckFromLua(L, 1);
+	const Color4ub *b = LuaColor::CheckFromLua(L, 2);
+	LuaColor::PushToLua(L, *a + *b);
+	return 1;
+}
+
+static int l_color_mul(lua_State *L)
+{
+	if (lua_isnumber(L, 1)) {
+		const double s = lua_tonumber(L, 1);
+		const Color4ub *c = LuaColor::CheckFromLua(L, 2);
+		LuaColor::PushToLua(L, *c * s);
+	} else if (lua_isnumber(L, 2)) {
+		const Color4ub *c = LuaColor::CheckFromLua(L, 1);
+		const double s = lua_tonumber(L, 2);
+		LuaColor::PushToLua(L, *c * s);
+	} else {
+		return luaL_error(L, "Cannot multiply a Color time Color");
```

Comment:
...Yes you can multiply two colors together. It's vector multiplication, you convert `0..255` to `0..1f`, multiply, and convert back to `0..255`.

Additionally, this function assumes that you're multiplying a color with a color - you could be using a boolean by accident. Please use `luaL_typename` to generate a correct error message.

Reply:
> ..Yes you can multiply two colors together. It's vector multiplication, you convert 0..255 to 0..1f, multiply, and convert back to 0..255.

...so use it at your own risk :wink:  

there's lines of code there:
https://github.com/pioneerspacesim/pioneer/blob/4c51f4eff542fde17bde94b9ccb6209da3a57f34/src/Color.h#L114

...probably another hint for future improvement, isn't?

> Additionally, this function assumes that you're multiplying a color with a color - you could be using a boolean by accident. Please use luaL_typename to generate a correct error message.

...I never seen so many checks around... This could be the first time, but then someone should check types for all Lua* types: sure a lot of big trees of if-then-else will kick in :smile: